### PR TITLE
Add documentations

### PR DESCRIPTION
**Pull request by yycdavid**
Add documentations and comments for the main struct and functions. The main parts I wrote are: src/model.rs:TensorAnalysis for creating metadata; src/rewrite.rs:CheckApply for checking before applying a pattern; src/optimize.rs:TensorCost for getting the cost during extraction; src/main.rs:optimize() main script, and convert_rw_rules() for converting TASO rules to proper format.

We can use this as a tryout for our review process.

### CODE REVIEW

File: src/model.rs [hunk 1]
```diff
@@ -53,47 +40,79 @@ define_language! {
...
+    fn default() -> Self {
+        DataKind::Name
+    }
 }
 
+/// Metadata struct for TensorAnalysis
+///
+/// # Fields
+///
+/// - `dtype`: The data type of this eclass, can be a name/scalar/tensor
```
Comment (mangpo) [inline src/model.rs:62]: Move these into the struct. Comment above each field.

Comment (mangpo) [inline src/model.rs]: Move these into the struct. Comment above each field.

File: src/model.rs [hunk 2]
```diff
@@ -53,47 +40,79 @@ define_language! {
...
 
+/// Struct for metadata analysis
+///
+/// In this analysis, it calls functions on the TASO side (e.g. graph.matmul())
+/// to create (or get) new ops/nodes and stores pointers to the output tensors.
+/// TASO will measure and store the runtime cost when creating a new op/node.
+///
+/// # Fields
+///
+/// - `graph`: Points to the graph object on the TASO side
```
Comment (mangpo) [inline src/model.rs]: Move inside struct. Same for the rest of the code.

**Review by mangpo — APPROVED** [review]

File: src/model.rs (hunk 1 above)
Comment (yycdavid) [reply src/model.rs]: I looked up some rust comment style guide, and seems that Rust convention is to have documentations above struct/function: https://deterministic.space/machine-readable-inline-markdown-code-cocumentation.html. Also checked egg: https://github.com/mwillsey/egg. Perhaps it's some special convention for Rust...

Comment (mangpo) [reply src/model.rs]: I think the template you use is for function. See: https://doc.rust-lang.org/stable/rust-by-example/meta/doc.html

Comment (yycdavid) [reply src/model.rs]: I see! Just added a commit that changes it. Thanks!

**Pull request merged**